src/sa.py

```diff
@@ -1,15 +1,149 @@
 import argparse
+import fasta, fastq
+import re
 
+def getSuffixes(x: memoryview):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    return suffixes
+
+def getTrailingNumber(s):
+    m = re.search(r'\d+$', s)
+    return (s[:m.start()], int(s[m.start():]))
+
+def strcmp(s1, index, x):
+    max_i = min(len(s1), len(x)-index)
+    i = 0
+    while i < max_i and s1[i] == x[index+i]:
+        i += 1
+        if i >= max_i:
+            return 0
+    if s1[i] > x[index+i]: return 1
+    else: return -1
+
+
+def search(sa, pattern, genome):
+    low = 0
+    high = len(sa)-1
+
+    if low == high:
+        cmp = strcmp(pattern, sa[low], genome)
+        if cmp == 0:
+            yield sa[low] + 1
+            return
```

Comment:
Now _if_ you excluded `high` from the interval, `low == high` meant you didn't have any matches, and you avoided the special case here.

Why the `+ 1`, by the way? If `low` points to the first match (and it is a match when `cmp == 0`), should you return `sa[low]`? I'm confused, so I'll continue reading...